csrc/ir/utils.cpp

```diff
@@ -1310,4 +1310,121 @@ MmaOpDetails getMmaOpDetails(
+    const DataType expected_input_dtype) {
+  const Val* in_a = mma_op->inA();
+  const Val* in_b = mma_op->inB();
+
+  const auto tv_a = in_a->as<TensorView>();
+  const auto tv_b = in_b->as<TensorView>();
+
+  NVF_ERROR(
+      tv_a->nDims() == tv_b->nDims(),
+      "Either both or none of A and B should be batch");
+  // Verify that the broadcasted size is 3.
+  NVF_ERROR(
+      tv_a->nDims() == 3,
+      "MmaOp::evaluate is not implemented for size: ",
+      tv_a->nDims());
+
+  NVF_ERROR(
+      in_a->definition() != nullptr && in_a->definition()->isA<BroadcastOp>(),
+      "Currently, MmaOp::evaluate assumes the preceding op to be a broadcast.");
+  NVF_ERROR(
+      in_b->definition() != nullptr && in_b->definition()->isA<BroadcastOp>(),
+      "Currently, MmaOp::evaluate assumes the preceding op to be a broadcast.");
+  NVF_ERROR(
+      tv_a->getRootDomain().back()->isBroadcast(),
+      "Expected last dimension to be broadcasted for first operand.");
+  NVF_ERROR(
+      tv_b->getRootDomain().front()->isBroadcast(),
+      "Expected first dimension to be broadcasted for second operand.");
+
+  // ATen preserves the dtype of MmaOp inputs whereas MmaOp generates float
+  // outputs. To preserve numerical equivalence and precision, the output of
+  // ATen matmul should be the same as MmaOp out `eventually`.
+  // See https://github.com/NVIDIA/Fuser/pull/1874#discussion_r1516991574
+  // Supported cases:
+  //  1. MmaOp->out() and MmaOp->input() are the same dtype.
+  //  2. MmaOp->out() is followed by a CastOp() to the MmaOp->input() dtype.
+  // NOTE: Currently MmaOp only accepts Half and BFloat16 so case (1) does not
+  // occur.
+
+  NVF_ERROR(
+      *(tv_a->getDataType()) == *(tv_b->getDataType()),
+      "MmaOp inputs should be of the same dtype.")
+  NVF_ERROR(
+      *(tv_a->getDataType()) == expected_input_dtype,
+      "MmaOp inputs should be the same dtype as the output dtype of the final castOp.");
+}
+
+bool matchMatmulCast(const UnaryOp* cast_op, Val*& mma_lhs, Val*& mma_rhs) {
+  if (auto* mma = dynamic_cast<MmaOp*>(cast_op->input(0)->definition())) {
+    MmaOpUtils::verifyMmaOpForEvaluation(
+        mma, cast_op->out()->getDataType().value());
+    mma_lhs = mma->inA();
+    mma_rhs = mma->inB();
+    return true;
+  }
+  return false;
+}
+
+bool matchMatmulBiasCast(
+    const UnaryOp* cast_op,
+    Val*& mma_lhs,
+    Val*& mma_rhs,
+    Val*& bias) {
+  if (auto* binary = dynamic_cast<BinaryOp*>(cast_op->input(0)->definition())) {
+    if (binary->getBinaryOpType() == BinaryOpType::Add) {
+      if (auto* mma = dynamic_cast<MmaOp*>(binary->input(0)->definition())) {
+        DataType final_out_dtype = cast_op->out()->getDataType().value();
+
+        // Verify assumptions for MmaOp hold. Assign the values to Mma operands.
+        MmaOpUtils::verifyMmaOpForEvaluation(mma, final_out_dtype);
+        mma_lhs = mma->inA();
+        mma_rhs = mma->inB();
+
+        Val* bcast_bias = binary->input(1);
```

Comment:
I think the code can simplify using dynamic_cast and reusing the same variable `bias`. The latter is a bit subjective because one could argue for SSA, but I tend to trade that for having to come up with different names for bias at different stages. 

For example,
```
bias = binary->input(1);
auto* bcast = dynamic_cast<BroadCastOp*>(bias->definition());  // btw, dynamic_cast a nullptr returns a nullptr. 
NVF_ERROR(bcast != nullptr, ...);
Check the last dimension is broadcasted;

bias = bcast->input(0);
auto* cast = dynamic_cast<UnaryOp*>(bias->definition());
check cast is not nullptr;
check cast is indeed of cast. 

bias = cast->input(0);

return true;
```
A slight downside is that bias may be non-null when match returns false, but callers tend to not read bias at all when the matching failed, so it's fine. If you really want to, you can add a comment saying these output references may be changed even when the matching failed.